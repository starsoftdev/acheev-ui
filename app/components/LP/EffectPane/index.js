// @flow

import React, { Component } from 'react';
import type { Map } from 'immutable';

import { Tabs, Tab } from 'components/Tabs';
import Table from 'components/Table';
import CustomProgressBar from 'components/ProgressBar';

type Props = {
  data: Map<*, *>,
  progressBar?: any,
  message?: string,
};
type State = {
  activeKey: number,
};

const items = {
  symptomsHelped: {
    tabLabel: 'Symptoms Helped',
    thLabel: 'Symptom',
  },
  positiveEffects: {
    tabLabel: 'Positive Effects',
    thLabel: 'Effect',
  },
  negativeEffects: {
    tabLabel: 'Negative Effects',
    thLabel: 'Effect',
  },
};

class LPEffectPane extends Component<Props, State> {
  state = {
    activeKey: 0,
  };

  handleSelect = (activeKey: number) => {
    this.setState({ activeKey });
  };

  render() {
    const {
      data,
      progressBar: ProgressBar = CustomProgressBar,
      message = 'No data for selected period',
    } = this.props;
    if (!data || !data.size) return null;
    const { activeKey } = this.state;
    const keys = Object.keys(items);
    const noData = keys.every(i => !data.get(i).size);
    if (noData) return <div>{message}</div>;
    return (
      <Tabs activeKey={activeKey} onSelect={this.handleSelect}>
        {data.entrySeq().map(([i, item], index) => {
          if (item.size) {
            return (
              <Tab key={i} eventKey={index} title={items[i].tabLabel}>
                <Table>
                  <thead>
                    <tr>
                      <th>{items[i].thLabel}</th>
                      <th>Score</th>
                      <th>Top Product</th>
                    </tr>
                  </thead>
                  <tbody>
                    {item.entrySeq().map(([subI, subItem]) => (
                      <tr key={subI}>
                        <td>{subItem.get('name')}</td>
                        <td>
                          <ProgressBar
                            value={subItem.get('score')}
                            maxValue={10}
                            showValue
                            valuePostfix="/10"
                            width={11}
                          />
                        </td>
                        <td>{subItem.getIn(['product', 'name'])}</td>
                      </tr>
                    ))}
                  </tbody>
                </Table>
              </Tab>
            );
          }
          return null;
        })}
      </Tabs>
    );
  }
}

export default LPEffectPane;
