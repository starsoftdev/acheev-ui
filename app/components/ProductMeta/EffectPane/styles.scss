@import '~styles/settings/settings';

.effectPane {
  .tabs-content {
    @include breakpoint(large up) {
      min-height: rem-calc(310);
    }
  }
}


