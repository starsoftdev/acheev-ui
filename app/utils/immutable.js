// @flow

import { Iterable } from 'immutable';
import type { Map } from 'immutable';
import { get as _get } from 'lodash-es';

export const get = (
  collection: ?Map<*, *> | Object,
  path: string | number,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.get(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};

export const getIn = (
  collection: ?Map<*, *> | Object,
  path: Array<string | number>,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.getIn(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};
