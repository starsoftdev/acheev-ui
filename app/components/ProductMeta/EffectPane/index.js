// @flow

import React, { Component } from 'react';
import { Tabs, Tab } from 'components/Tabs';
import EffectPanel from '../EffectPanel';

import './styles.scss';

type Props = {
  data: Object,
};
type State = {
  activeKey: string,
};

class EffectPane extends Component<Props, State> {
  state = {
    activeKey: '1',
  };

  handleSelect = (activeKey: string) => {
    this.setState({ activeKey });
  };

  render() {
    const { data } = this.props;
    const { activeKey } = this.state;
    const symptomsHelped = data.get('symptomsHelped');
    const positiveEffects = data.get('positiveEffects');
    const negativeEffects = data.get('negativeEffects');
    return (
      <div className="effectPane">
        <Tabs activeKey={activeKey} onSelect={this.handleSelect}>
          <Tab eventKey="1" title="Symptoms Helped">
            {symptomsHelped && symptomsHelped.size > 0 ? (
              <EffectPanel data={symptomsHelped} />
            ) : (
              'No data available'
            )}
          </Tab>
          <Tab eventKey="2" title="Positive Effects">
            {positiveEffects && positiveEffects.size > 0 ? (
              <EffectPanel data={positiveEffects} />
            ) : (
              'No data available'
            )}
          </Tab>
          <Tab eventKey="3" title="Negative Effects">
            {negativeEffects && negativeEffects.size > 0 ? (
              <EffectPanel data={negativeEffects} />
            ) : (
              'No data available'
            )}
          </Tab>
        </Tabs>
      </div>
    );
  }
}

export default EffectPane;
