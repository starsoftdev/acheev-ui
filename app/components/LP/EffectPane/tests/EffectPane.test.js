import React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';
import { fromJS } from 'immutable';

import EffectPane from 'components/LP/EffectPane';

const data = fromJS({
  symptomsHelped: [
    {
      name: 'anxiety',
      score: '5.0',
      product: {
        _id: '5a39d75968ab9f3e45b98d89',
        slug: 'green-relief-low-tide-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Low Tide',
        __t: 'Strain',
      },
    },
    {
      name: 'back pain',
      score: '2.5',
      product: {
        _id: '5a39d50668ab9f3e45b98d71',
        slug: 'green-relief-tsunami-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Tsunami',
        __t: 'Strain',
      },
    },
    {
      name: 'appetite',
      score: '2.0',
      product: {
        _id: '5a39d75968ab9f3e45b98d89',
        slug: 'green-relief-low-tide-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Low Tide',
        __t: 'Strain',
      },
    },
  ],
  positiveEffects: [
    {
      name: 'anti-anxiety',
      score: '5.0',
      product: {
        _id: '5a39d75968ab9f3e45b98d89',
        slug: 'green-relief-low-tide-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Low Tide',
        __t: 'Strain',
      },
    },
    {
      name: 'appetite',
      score: '5.0',
      product: {
        _id: '5a39d50668ab9f3e45b98d71',
        slug: 'green-relief-tsunami-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Tsunami',
        __t: 'Strain',
      },
    },
    {
      name: 'sleepy',
      score: '1.0',
      product: {
        _id: '5a39d53468ab9f3e45b98d73',
        slug: 'green-relief-ripple-effect-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Ripple Effect',
        __t: 'Strain',
      },
    },
  ],
  negativeEffects: [
    {
      name: 'harsh',
      score: '7.0',
      product: {
        _id: '5a39d53468ab9f3e45b98d73',
        slug: 'green-relief-ripple-effect-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Ripple Effect',
        __t: 'Strain',
      },
    },
    {
      name: 'chest congestion',
      score: '7.0',
      product: {
        _id: '5a39d53468ab9f3e45b98d73',
        slug: 'green-relief-ripple-effect-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Ripple Effect',
        __t: 'Strain',
      },
    },
    {
      name: 'coughing',
      score: '6.0',
      product: {
        _id: '5a39d53468ab9f3e45b98d73',
        slug: 'green-relief-ripple-effect-strain',
        business: '58c0bdd4bdabcfc75b4112a4',
        name: 'Ripple Effect',
        __t: 'Strain',
      },
    },
  ],
});

const emptyData = fromJS({
  negativeEffects: [],
  positiveEffects: [],
  symptomsHelped: [],
});

describe('<EffectPane />', () => {
  it("shouldn't throw error if data is not specified", () => {
    try {
      shallow(<EffectPane />);
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });

  it('should render `message` if empty data is provided', () => {
    const renderedComponent = shallow(
      <EffectPane data={emptyData} message="Hey" />
    );
    expect(renderedComponent.contains('Hey')).toEqual(true);
  });

  it('should match snapshoot', () => {
    const component = create(
      <EffectPane data={data} progressBar="ProgressBar" />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
