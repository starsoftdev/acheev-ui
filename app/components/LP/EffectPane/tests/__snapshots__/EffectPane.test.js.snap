// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EffectPane /> should match snapshoot 1`] = `
<div>
  <div
    className="tabs"
  >
    <li
      className="tabs-title is-active"
      role="presentation"
    >
      <a
        aria-selected={true}
        href="#"
        id={null}
        onClick={[Function]}
        role="tab"
        title="Symptoms Helped"
      >
        Symptoms Helped
      </a>
    </li>
    <li
      className="tabs-title"
      role="presentation"
    >
      <a
        aria-selected={false}
        href="#"
        id={null}
        onClick={[Function]}
        role="tab"
        title="Positive Effects"
      >
        Positive Effects
      </a>
    </li>
    <li
      className="tabs-title"
      role="presentation"
    >
      <a
        aria-selected={false}
        href="#"
        id={null}
        onClick={[Function]}
        role="tab"
        title="Negative Effects"
      >
        Negative Effects
      </a>
    </li>
  </div>
  <div
    className="tabs-content"
  >
    <div
      aria-hidden={false}
      aria-labelledby={null}
      className="tabs-panel is-active"
      role="tabpanel"
      title="Symptoms Helped"
    >
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              Symptom
            </th>
            <th>
              Score
            </th>
            <th>
              Top Product
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              anxiety
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="5.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Low Tide
            </td>
          </tr>
          <tr>
            <td>
              back pain
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="2.5"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Tsunami
            </td>
          </tr>
          <tr>
            <td>
              appetite
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="2.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Low Tide
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="tabs-panel"
      role="tabpanel"
      title="Positive Effects"
    >
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              Effect
            </th>
            <th>
              Score
            </th>
            <th>
              Top Product
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              anti-anxiety
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="5.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Low Tide
            </td>
          </tr>
          <tr>
            <td>
              appetite
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="5.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Tsunami
            </td>
          </tr>
          <tr>
            <td>
              sleepy
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="1.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Ripple Effect
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="tabs-panel"
      role="tabpanel"
      title="Negative Effects"
    >
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              Effect
            </th>
            <th>
              Score
            </th>
            <th>
              Top Product
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              harsh
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="7.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Ripple Effect
            </td>
          </tr>
          <tr>
            <td>
              chest congestion
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="7.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Ripple Effect
            </td>
          </tr>
          <tr>
            <td>
              coughing
            </td>
            <td>
              <ProgressBar
                maxValue={10}
                showValue={true}
                value="6.0"
                valuePostfix="/10"
                width={11}
              />
            </td>
            <td>
              Ripple Effect
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
